tests/checks/preflight.go

```diff
@@ -0,0 +1,139 @@
+	// kubernetesVersion specifies the kubernetes version of the test cluster
+	// All tests which which require <= this version are supposed to run.
+	KubernetesVersion *string `json:"kubernetesVersion,omitempty"`
+	// dualNetworkStack indicates if the cluster provides IPv4 and IPv6 addresses
+	// All tests which require a dual-stack setup are supposed to run.
+	DualNetworkStack *bool `json:"dualNetworkStack,omitempty"`
+	// minimumSchedulableNodes is the minimum of nodes which are capable of scheduling and running VMs
+	// All tests which require <= this amount of nodes are supposed to be run
+	MinimumSchedulableNodes *int `json:"minimumSchedulableNodes,omitempty"`
+	// nodesWithCPUManager contains the number of nodes with the CPU manager preset
+	MinimumNodesWithCPUManager *int `json:"minimumNodesWithCPUManager,omitempty"`
+	// isKind tells if the tests are running on kind
+	IsKind *bool `json:"isKind,omitempty"`
+}
+
+func IntrospectCluster() {
+	virtClient, err := kubecli.GetKubevirtClient()
+	util.PanicOnError(err)
+	nodes := util.GetAllSchedulableNodes(virtClient).Items
+	nodeCount := len(nodes)
+	DiscoveredClusterProfile.MinimumSchedulableNodes = &nodeCount
+	if IsOpenShift() {
+		majorVersion := cluster.GetOpenShiftMajorVersion(virtClient)
+		DiscoveredClusterProfile.OpenShiftMajorVersion = &majorVersion
+	}
+	kubernetesVersion, err := cluster.GetKubernetesVersion(virtClient)
+	if err != nil {
+		util.PanicOnError(err)
+	}
+	DiscoveredClusterProfile.KubernetesVersion = &kubernetesVersion
+
+	dualStack, err := libnet.IsClusterDualStack(virtClient)
+	if err != nil {
+		util.PanicOnError(err)
+	}
+	DiscoveredClusterProfile.DualNetworkStack = &dualStack
+
+	withCPUManager := 0
+	for _, node := range nodes {
+		if IsCPUManagerPresent(&node) {
+			withCPUManager++
+		}
+	}
+	DiscoveredClusterProfile.MinimumNodesWithCPUManager = &withCPUManager
+
+	isKind := IsRunningOnKindInfra() || IsRunningOnKindInfraIPv6()
+	DiscoveredClusterProfile.IsKind = &isKind
+
+	encoder := json.NewEncoder(os.Stdout)
+	encoder.SetIndent("", "  ")
+	err = encoder.Encode(&DiscoveredClusterProfile)
+	util.PanicOnError(err)
+}
+
+type DiscoveryError struct {
+	errors []error
+}
+
+func (d *DiscoveryError) Error() string {
+	return fmt.Sprintf("%v", d.errors)
+}
+
+func VerifyClusterExpectations() error {
+	if len(flags.ClusterProfilePath) == 0 {
+		return nil
+	}
+
+	expected := &ClusterProfile{}
+	reader, err := os.Open(flags.ClusterProfilePath)
+	if err != nil {
+		log.DefaultLogger().Reason(err).Critical("Could not find the provided cluster profile file.")
+	}
+	if err := yaml.NewYAMLOrJSONDecoder(reader, 1024).Decode(expected); err != nil {
+		log.DefaultLogger().Reason(err).Critical("Could not decode the provided cluster profile file.")
+	}
+
+	errors := &DiscoveryError{}
+	if expected.OpenShiftMajorVersion != nil {
+		if !reflect.DeepEqual(*expected.OpenShiftMajorVersion, *DiscoveredClusterProfile.OpenShiftMajorVersion) {
```

Comment:
The method assumes `DiscoveredClusterProfile` was already initialized. I would change `IntrospectCluster` to be a private method and invoke at the beginning of `IntrospectCluster`,